hyperdash_cli/cli.py

```diff
@@ -128,6 +134,85 @@ def demo(args=None):
+    access_token_queue = Queue()
+
+    # Server that we will run in the background to accept a post-OAuth redirect from
+    # the Hyperdash server which will contain the user's access token
+    def start_server():
+        class OAuthRedirectHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_GET(self):
+                parsed_path = urlparse(self.path)
+                query = parse_qs(parsed_path.query)
+                access_token = query["access_token"][0]
+                access_token_queue.put(access_token)
+                # Redirect user's browser
+                self.send_response(301)
+                self.send_header('Location',"{}/{}".format(get_base_http_url(), "/oauth/github/success"))
+                self.end_headers()
+
+        server = BaseHTTPServer.HTTPServer(('localhost', port), OAuthRedirectHandler)
+        server_started_queue.put(True)
+        server.handle_request()
+    
+    server_thread = Thread(target=start_server)
+    # Prevent server_thread from preventing program shutdown
+    server_thread.setDaemon(True)
+    server_thread.start()
+    # Wait until the server has started before opening the user's browser to prevent
+    # a race condition where the Hyperdash server redirects with an access token but
+    # the Python server isn't read yet
+    print("Opening browser, please wait. If something goes wrong, press CTRL+C to cancel.")
+    server_started_queue.get(block=True)
+    # Blocks until browser opens, but doesn't wait for user to close it
+    # TODO: Constantize this
+    webbrowser.open_new_tab("https://github.com/login/oauth/authorize?client_id=4793c654216adde7a478&redirect_uri=http://localhost:4040/oauth/github/callback&allow_signup=false&scope=user%20email&state=client_cli:{}".format(port))
+
+
+    print("Waiting for Github OAuth to complete. If something goes wrong, press CTRL+C to cancel.")
+    # Wait for the Hyperdash server to redirect with the access token or an error
+    access_token = access_token_queue.get(block=True)
+    # Use the access token to retrieve the user's API key and store a valid
+    # hyperdash.json file
+    success, default_api_key = _after_access_token_login(access_token)
+    if success:
+        print("Successfully logged in! We also installed: {} as your default API key".format(
+            default_api_key))
+
+
+def _find_available_port():
+    port = None
+    for cur_port in xrange(3000, 3100):
+        is_open = _is_port_open("127.0.0.1", cur_port)
+        if is_open:
+            port = cur_port
+            break
+    return port
+
+
+def _is_port_open(host, port):
+    try:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.bind(('127.0.0.1', port))
+        sock.listen(5)
+        sock.close()
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
+        sock.bind(('::1', port))
+        sock.listen(5)
+        sock.close()
+    except socket.error as e:
+        return False
```

Comment:
Looked into error handling for `socket`. Turns out there are a few exception types beyond `socket.error`. Doesn't seem like we're probable to trigger the others with the `_is_port_open` logic, but I'd recommend a more generic error catcher to be safe.

> https://docs.python.org/2/library/socket.html
> exception socket.error
> This exception is raised for socket-related errors. The accompanying value is either a string telling what went wrong or a pair (errno, string) representing an error returned by a system call, similar to the value accompanying os.error. See the module errno, which contains names for the error codes defined by the underlying operating system.
> 
> exception socket.herror
> This exception is raised for address-related errors, i.e. for functions that use h_errno in the C API, including gethostbyname_ex() and gethostbyaddr().
> 
> The accompanying value is a pair (h_errno, string) representing an error returned by a library call. string represents the description of h_errno, as returned by the hstrerror() C function.
> 
> exception socket.gaierror
> This exception is raised for address-related errors, for getaddrinfo() and getnameinfo(). The accompanying value is a pair (error, string) representing an error returned by a library call. string represents the description of error, as returned by the gai_strerror() C function. The error value will match one of the EAI_* constants defined in this module.
> 
> exception socket.timeout
> This exception is raised when a timeout occurs on a socket which has had timeouts enabled via a prior call to settimeout(). The accompanying value is a string whose value is currently always “timed out”.

Reply:
for socket.herror we're not using any of the address-related functions, same for socket.gaierror. Good call on the timeout one, I meant to add a timeout but forgot